src/server/scripts/Outland/BlackTemple/boss_shade_of_akama.cpp

```diff
@@ -157,236 +210,151 @@ class boss_shade_of_akama : public CreatureScript
             {
-                combatStarted = true;
-                events.ScheduleEvent(EVENT_START_ATTACK_AKAMA, 500);
-                events.ScheduleEvent(EVENT_SET_CHANNELERS_SPAWNERS, 1000);
+                events.ScheduleEvent(EVENT_START_CHANNELERS_AND_SPAWNERS, Seconds(1));
                 me->SetUInt32Value(UNIT_NPC_EMOTESTATE, EMOTE_STATE_NONE);
+                events.ScheduleEvent(EVENT_EVADE_CHECK, Seconds(10));
                 if (Creature* Akama = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_AKAMA_SHADE)))
-                    me->AddThreat(Akama, 10000000.0f);
+                {
+                    AttackStart(Akama);
+                    me->AddThreat(Akama, 100000.0f);
+                }
             }
-            else if (spell->Id == SPELL_SHADE_SOUL_CHANNEL_2)
-                UpdateSpeed();
+
+            if (spell->Id == SPELL_AKAMA_SOUL_EXPEL)
+                DoCastSelf(SPELL_AKAMA_SOUL_EXPEL_CHANNEL);
         }
 
-        void UpdateSpeed()
+        void MoveInLineOfSight(Unit* who) override
         {
-            float moveSpeed = 0.2f;
-
-            if (me->GetAuraCount(SPELL_SHADE_SOUL_CHANNEL_2) <= 3)
+            if (who->GetGUID() == instance->GetGuidData(DATA_AKAMA_SHADE) && me->IsWithinMeleeRange(who))
             {
-                moveSpeed = (2.0f - (0.6f * me->GetAuraCount(SPELL_SHADE_SOUL_CHANNEL_2)));
-                me->SetSpeedRate(MOVE_WALK, moveSpeed / 2.5f);
-                me->SetSpeedRate(MOVE_RUN, (moveSpeed * 2) / 7);
-                me->ClearUnitState(UNIT_STATE_ROOT);
+                me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NOT_SELECTABLE);
+                me->SetWalk(false);
+                events.ScheduleEvent(EVENT_ADD_THREAT, Milliseconds(100));
+
+                for (GuidVector::const_iterator itr = Spawners.begin(); itr != Spawners.end(); ++itr)
+                    if (Creature* Spawner = ObjectAccessor::GetCreature(*me, *itr))
+                        Spawner->AI()->DoAction(ACTION_STOP_SPAWNING);
             }
-            else
-                me->AddUnitState(UNIT_STATE_ROOT);
         }
 
-        void UpdateAI(uint32 diff) override
+        void JustDied(Unit* /*killer*/) override
         {
-            if (HasKilledAkamaAndReseting)
-                return;
+            DoCastSelf(SPELL_SHADE_OF_AKAMA_TRIGGER);
+
+            if (Creature* Akama = ObjectAccessor::GetCreature(*me, instance->GetGuidData(DATA_AKAMA_SHADE)))
+                Akama->AI()->DoAction(ACTION_SHADE_OF_AKAMA_DEAD);
+
+            for (GuidVector::const_iterator itr = Spawners.begin(); itr != Spawners.end(); ++itr)
+                if (Creature* Spawner = ObjectAccessor::GetCreature(*me, *itr))
+                    Spawner->AI()->DoAction(ACTION_DESPAWN_ALL_SPAWNS);
+
+            events.Reset();
+            summons.DespawnEntry(NPC_ASHTONGUE_CHANNELER);
+            instance->SetBossState(DATA_SHADE_OF_AKAMA, DONE);
+        }
+
+        void CheckIfNeedsToEvade()
```

Comment:
Having a method called "Check" do something is bad form, imo.

Either return bool and check, then evade in caller; or rename it.